### PR TITLE
fix(terraform): Replace correct currentValue match in constraint with new value

### PR DESCRIPTION
**Pull request by dali546**
## Changes

Terraform constraints has the `currentVersion` multiple times in the string. However only the `required_provider` currentValue should be updated, instead of the first occurrence of `currentValue`

eg. Terraform hashicorp/aws provider update didn't update the last occurrence of `5.31.0`

```diff
provider "registry.opentofu.org/hashicorp/aws" {
-   version     = "5.31.0"
-   constraints = ">= 2.49.0, >= 3.29.0, >= 3.75.0, >= 4.0.0, >= 4.9.0, >= 4.40.0, >= 4.42.0, >= 4.66.1, >= 5.26.0, >= 5.31.0, 5.31.0"
+   version     = "5.34.0"
+   constraints = ">= 2.49.0, >= 3.29.0, >= 3.75.0, >= 4.0.0, >= 4.9.0, >= 4.40.0, >= 4.42.0, >= 4.66.1, >= 5.26.0, >= 5.34.0, 5.31.0"
```

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
Please add a test

**Review by secustor — COMMENTED** [review]
What is your reasoning that you want to have changed multiple ranges?

Terraform does not use this field and we try to keep the diff to a minimum.

**Comment by dali546** [comment]
@secustor The lockfile example I gave is an extract of a lockfile that Terraform generated. The terraform workspace has required providers, but also many modules in use, which also define their own provider constraints.

`terraform init` (or `tofu init`) will generate a lockfile with constraints like this.

**Comment by dali546** [comment]
@secustor 

Here is a minimal terraform file to reproduce
```hcl
terraform {
  required_version = "~> 1.6.0"
  required_providers {
    name = {
      source  = "hashicorp/aws"
      version = "5.31.0"
    }
  }
}

module "nlb" {
  source  = "terraform-aws-modules/alb/aws"
  version = "9.4.1"
}
```
This will produce a lockfile with multiple ranges

**Comment by secustor** [comment]
I'm aware of that tough we can not know what the constraints of the modules are. 

Therefore we only look at the constraint of the `required_provider` block, but as I said this is only cosmetic as Terraform does not use the constraints field.

**Comment by dali546** [comment]
@secustor Terraform does indeed look at the `constraints` field when doing `terraform init`. Our CI checks fail because the Terraform dependency updates cannot find a suitable provider version to install

**Comment by dali546** [comment]
@secustor How about I fix this then. Instead of matching the first match of the version, (in this case there is `>= 4.31.0` and `4.31.0`

The second `4.31.0` constraint is set by the `required_provider` ... Is it better to split by `,` and replace the string exact match?

**Comment by secustor** [comment]
Your terraform init error is probably not coming from the lock file but rather because the module you are depending on is requesting this. 

Take look at this discussion for some context: https://github.com/renovatebot/renovate/discussions/24879#discussioncomment-7261513

`@apparentlymart`, a maintainer of Terraform give his insight there too. 

Further please do not ping maintainers, this is considered to be rude.

**Comment by dali546** [comment]
https://github.com/dali546/terraform-renovate-bug

Apologies for the mentioning you, Here is a repo reproducing the issue. Take a look at the CI checks to show Simple terraform pipeline failing

**Comment by secustor** [comment]
Looking at the error message, it rather looks like that Terraform expects now that this string is order in ascending order of the version 🤔 

It also writes out the string it expects:
https://github.com/dali546/terraform-renovate-bug/actions/runs/7695970828/job/20969958070?pr=1#step:4:14

**Comment by dali546** [comment]
I have manually updated it locally, and the order of the constraints doesn't matter (it doesn't need to be in ascending order). As long as the constraint it valid. 

https://github.com/dali546/terraform-renovate-bug/pull/4

File: lib/modules/manager/terraform/lockfile/index.spec.ts [hunk 1]
```diff
@@ -1118,6 +1118,19 @@ describe('modules/manager/terraform/lockfile/index', () => {
...
+        getNewConstraint(
+          {
+            currentValue: '2.41.0',
+            newValue: '2.46.0',
+            newVersion: '2.46.0',
+          },
+          '>= 2.41.0, 2.41.0',
+        ),
+      ).toBe('>= 2.41.0, 2.46.0');
+    });
```
Comment (secustor) [inline lib/modules/manager/terraform/lockfile/index.spec.ts:1132]: Add an additional test where the fixed constraint is in the middle.

**Review by secustor — COMMENTED** [review]
Else lgtm

File: lib/modules/manager/terraform/lockfile/index.spec.ts [hunk 2]
```diff
@@ -1118,6 +1118,32 @@ describe('modules/manager/terraform/lockfile/index', () => {
...
+      expect(
+        getNewConstraint(
+          {
+            currentValue: '<= 2.41.0',
+            newValue: '<= 2.46.0',
+            newVersion: '2.46.0',
+          },
+          '>= 2.41.0, <= 2.41.0, >= 2',
+        ),
+      ).toBe('>= 2.41.0, <= 2.46.0, >= 2');
```
Comment (secustor) [inline lib/modules/manager/terraform/lockfile/index.spec.ts:1144]:
```suggestion
          '>= 2.41.0, <= 2.41.0, >= 2.0.0',
        ),
      ).toBe('>= 2.41.0, <= 2.46.0, >= 2.0.0');
```
Versions in Terraform lockfiles have to be full semver versions

**Review by secustor — APPROVED** [review]

**Comment by dali546** [comment]
@rarkins can you review this PR so it can get merged?

**Review by rarkins — APPROVED** [review]

**Pull request merged**